### ISSUE TITLE
Too many threads with asyn client

### ISSUE THREAD

**Issue by pengweichu** [Corvusoft/restbed#73]
I've tried the http client in async mode, but it's seems has a thread problem, after a few requests completed, there remains too many threads.

Below is my code and I have set the break point at last line: return EXIT_SUCCESS; then watch the threads, please see attached the screenshot of threads.
``` C++
#include <Windows.h>
#include <memory>
#include <cstdio>
#include <cstdlib>
#include <restbed>
#include <functional>

using namespace std;
using namespace restbed;

void cb(const std::shared_ptr< Request > request, const std::shared_ptr< Response > response)
{
	fprintf(stderr, "*** Response ***\n");
	fprintf(stderr, "Status Code:    %i\n", response->get_status_code());
	fprintf(stderr, "Status Message: %s\n", response->get_status_message().data());
	fprintf(stderr, "HTTP Version:   %.1f\n", response->get_version());
	fprintf(stderr, "HTTP Protocol:  %s\n", response->get_protocol().data());

	Http::close(request);
	Http::close(response);
}

int main( const int, const char** )
{
	auto settings = make_shared<restbed::Settings>();
	settings->set_port(80);
	settings->set_case_insensitive_uris(false);
	settings->set_worker_limit(1);
	settings->set_connection_timeout(std::chrono::seconds(5));

	for (int i = 0; i < 10; ++ i)
	{
		Sleep(2000);

		auto request = make_shared< Request >(Uri("https://www.google.com/"));
		request->set_header("Accept", "*/*");
		request->set_header("Host", "www.google.com");
		request->set_query_parameter("query", "search term");

		auto response = Http::async(request, std::bind(cb, std::placeholders::_1, std::placeholders::_2), settings);
	}


	Sleep(1000 * 60);
    

    return EXIT_SUCCESS;
}
![threads](https://cloud.githubusercontent.com/assets/12697283/13074468/a98b197a-d4df-11e5-8913-51498d424ef5.png)
````

**Comment by ben-crowhurst** [comment]
**Code Review**
1. No need to perform the following.  Calling close on the request will close the response and vice-versa;  Alternatively when the request and response go out of scope, the socket will be closed.
``` C++
    Http::close(request);
    Http::close(response);
```

2. ```Sleep(2000)``` How is this implemented?  make sure you're using std::this_thread::sleep_for and not putting the entire process to rest.

3. The following settings are Service specific and have no effect on the Http client.
``` C++
    settings->set_port( 80 );
    settings->set_case_insensitive_uris( false );
    settings->set_worker_limit( 1 );
```

4. ```std::bind(cb, std::placeholders::_1, std::placeholders::_2)``` This is overly verbose, read std::bind documentation.

5. Your understanding of variable scope seems flawed.  You create request/response instances within a for-loop construct, which is reclaimed on the next iteration.  The following would remedy these short comings.

``` C++
    auto settings = make_shared< Settings >( );
    settings->set_connection_timeout( seconds( 5 ) );

    shared_ptr< Request > request;
    vector< future< shared_ptr< Response > > > responses;

    for ( int thread_count = 0; thread_count < 10; thread_count++ )
    {
        request = make_shared< Request >( Uri( "https://www.google.com/" ) );
        request->set_header( "Accept", "*/*" );
        request->set_header( "Host", "www.google.com" );
        request->set_query_parameter( "query", "search term" );

        responses.push_back( Http::async( request, completion_handler, settings) );
    }

    for ( auto& response : responses )
    {
    	fprintf( stderr, "Waiting on response future..." );
    	
    	response.wait( );
        Http::close( response.get( ) );

    	fprintf( stderr, "\tCompletion handler invoked.\n");
    }

    responses.clear( );
```

**Answer**
The restbed::Http client uses [std::async](http://en.cppreference.com/w/cpp/thread/async) internally.  No threads are directly instantiated by the client codebase.

Reading the manual for std::async we see that a thread can potentially be created on thread pool:

> The template function async runs the function f asynchronously (potentially in a separate thread which may be part of a thread pool) and returns a std::future that will eventually hold the result of that function call.

Thus the threads are hanging around to improve performance at a later date, if required.  You can validate this by inspecting each thread, you'll notice they invoke sleep.

When and how this thread pool is managed is implementation specific (Unix, Mac, Win, etc...).

**Recommendations**
* Learn about [C++ scoping rules](https://en.wikipedia.org/wiki/Scope_(computer_science).
* Learn about [http://en.cppreference.com/w/cpp/thread/async](std::async)
* Learn about [http://en.cppreference.com/w/cpp/thread/future](std::future)
* Learn about the [C++ Standard Template Library](https://en.wikipedia.org/wiki/Standard_Template_Library)

**Comment by pengweichu** [comment]
@ben-crowhurst The same result if I using your code:

void cb(const std::shared_ptr< Request > request, const std::shared_ptr< Response > response)
{
	fprintf(stderr, "*** Response ***\n");
}

int main( const int, const char** )
{

	auto settings = make_shared< Settings >();
	settings->set_connection_timeout(std::chrono::seconds(5));

	shared_ptr< Request > request;
	vector< future< shared_ptr< Response > > > responses;

	for (int thread_count = 0; thread_count < 10; thread_count++)
	{
		request = make_shared< Request >(Uri("https://www.google.com/"));
		request->set_header("Accept", "*/*");
		request->set_header("Host", "www.google.com");
		request->set_query_parameter("query", "search term");

		responses.push_back(Http::async(request, cb, settings));
	}

	for (auto& response : responses)
	{
		fprintf(stderr, "Waiting on response future...");

		response.wait();
		Http::close(response.get());

		fprintf(stderr, "\tCompletion handler invoked.\n");
	}

	responses.clear();


	std::this_thread::sleep_for(std::chrono::seconds(60));
    

    return EXIT_SUCCESS;
}
![t2](https://cloud.githubusercontent.com/assets/12697283/13097354/6067337e-d55b-11e5-8a8d-8a09bed35631.png)

**Comment by ben-crowhurst** [comment]
Please carefully retread my previous response.
